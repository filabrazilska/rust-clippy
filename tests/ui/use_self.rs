

#![warn(use_self)]
#![allow(dead_code)]
#![allow(should_implement_trait)]
#![allow(plugin_as_library)]

#[macro_use]
extern crate clippy_mini_macro_test;

fn main() {}

mod use_self {
    struct Foo {}

    impl Foo {
        fn new() -> Foo {
            Foo {}
        }
        fn test() -> Foo {
            Foo::new()
        }
    }

    impl Default for Foo {
        fn default() -> Foo {
            Foo::new()
        }
    }
}

mod internal_macro_fails {
    macro_rules! expand_wrong_functions {
        () => (
            fn new() -> FooWrong {
                FooWrong {}
            }
            fn test() -> FooWrong {
                FooWrong::new()
            }
        )
    }

    macro_rules! expand_better_functions {
        () => (
            fn new() -> Self {
                Self {}
            }
            fn test() -> Self {
                Self::new()
            }
        )
    }

    struct FooWrong {}

    impl FooWrong {
        expand_wrong_functions!();
    }

    struct FooBetter {}

    impl FooBetter {
        expand_better_functions!();
    }
}

mod external_macro_not_linted {
    struct FooWrong {}

    impl FooWrong {
        use_self_expand_wrong_functions!();
    }
}

mod better {
    struct Foo {}

    impl Foo {
        fn new() -> Self {
            Self {}
        }
        fn test() -> Self {
            Self::new()
        }
    }

    impl Default for Foo {
        fn default() -> Self {
            Self::new()
        }
    }
}

// todo the lint does not handle lifetimed struct
// the following module should trigger the lint on the third method only
mod lifetimes {
    struct Foo<'a> {
        foo_str: &'a str,
    }

    impl<'a> Foo<'a> {
        // Cannot use `Self` as return type, because the function is actually `fn foo<'b>(s: &'b str) ->
        // Foo<'b>`
        fn foo(s: &str) -> Foo {
            Foo { foo_str: s }
        }
        // cannot replace with `Self`, because that's `Foo<'a>`
        fn bar() -> Foo<'static> {
            Foo { foo_str: "foo" }
        }

        // `Self` is applicable here
        fn clone(&self) -> Foo<'a> {
            Foo { foo_str: self.foo_str }
        }
    }
}
