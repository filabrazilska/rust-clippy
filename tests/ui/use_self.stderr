error: unnecessary structure name repetition
  --> $DIR/use_self.rs:17:21
   |
17 |         fn new() -> Foo {
   |                     ^^^ help: use the applicable keyword: `Self`
   |
   = note: `-D use-self` implied by `-D warnings`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:18:13
   |
18 |             Foo {}
   |             ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:20:22
   |
20 |         fn test() -> Foo {
   |                      ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:21:13
   |
21 |             Foo::new()
   |             ^^^^^^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:26:25
   |
26 |         fn default() -> Foo {
   |                         ^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:27:13
   |
27 |             Foo::new()
   |             ^^^^^^^^ help: use the applicable keyword: `Self`

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:35:25
   |
35 |             fn new() -> FooWrong {
   |                         ^^^^^^^^ help: use the applicable keyword: `Self`
...
58 |         expand_wrong_functions!();
   |         -------------------------- in this macro invocation

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:36:17
   |
36 |                 FooWrong {}
   |                 ^^^^^^^^ help: use the applicable keyword: `Self`
...
58 |         expand_wrong_functions!();
   |         -------------------------- in this macro invocation

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:38:26
   |
38 |             fn test() -> FooWrong {
   |                          ^^^^^^^^ help: use the applicable keyword: `Self`
...
58 |         expand_wrong_functions!();
   |         -------------------------- in this macro invocation

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:39:17
   |
39 |                 FooWrong::new()
   |                 ^^^^^^^^^^^^^ help: use the applicable keyword: `Self`
...
58 |         expand_wrong_functions!();
   |         -------------------------- in this macro invocation

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:72:9
   |
72 |         use_self_expand_wrong_functions!();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = note: this error originates in a macro outside of the current crate

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:72:9
   |
72 |         use_self_expand_wrong_functions!();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = note: this error originates in a macro outside of the current crate

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:72:9
   |
72 |         use_self_expand_wrong_functions!();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = note: this error originates in a macro outside of the current crate

error: unnecessary structure name repetition
  --> $DIR/use_self.rs:72:9
   |
72 |         use_self_expand_wrong_functions!();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = note: this error originates in a macro outside of the current crate

